src/number.rs

```diff
@@ -292,90 +381,355 @@ impl<'de> Deserialize<'de> for Number {
             {
                 Number::from_f64(value).ok_or_else(|| de::Error::custom("not a JSON number"))
             }
+
+            #[cfg(feature = "arbitrary_precision")]
+            #[inline]
+            fn visit_map<V>(self, mut visitor: V) -> Result<Number, V::Error>
+            where
+                V: de::MapAccess<'de>
+            {
+                let value = visitor.next_key::<NumberKey>()?;
+                if value.is_none() {
+                    return Err(de::Error::invalid_type(Unexpected::Map, &self));
+                }
+                let v: NumberFromString = visitor.next_value()?;
+                Ok(v.value)
+            }
         }
 
         deserializer.deserialize_any(NumberVisitor)
     }
 }
 
-impl<'de> Deserializer<'de> for Number {
-    type Error = Error;
+#[cfg(feature = "arbitrary_precision")]
+struct NumberKey;
 
-    #[inline]
-    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> de::Deserialize<'de> for NumberKey {
+    fn deserialize<D>(deserializer: D) -> Result<NumberKey, D::Error>
     where
-        V: Visitor<'de>,
+        D: de::Deserializer<'de>
     {
-        match self.n {
-            N::PosInt(i) => visitor.visit_u64(i),
-            N::NegInt(i) => visitor.visit_i64(i),
-            N::Float(f) => visitor.visit_f64(f),
+        struct FieldVisitor;
+
+        impl<'de> de::Visitor<'de> for FieldVisitor {
+            type Value = ();
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a valid number field")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<(), E>
+                where E: de::Error
+            {
+                if s == SERDE_STRUCT_FIELD_NAME {
+                    Ok(())
+                } else {
+                    Err(de::Error::custom("expected field with custom name"))
+                }
+            }
+        }
+
+        deserializer.deserialize_identifier(FieldVisitor)?;
+        Ok(NumberKey)
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+pub struct NumberFromString {
+    pub value: Number,
+}
+
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> de::Deserialize<'de> for NumberFromString {
+    fn deserialize<D>(deserializer: D) -> Result<NumberFromString, D::Error>
+    where
+        D: de::Deserializer<'de>
+    {
+        struct Visitor;
+
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = NumberFromString;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string containing a number")
+            }
+
+            fn visit_string<E>(self, s: String) -> Result<NumberFromString, E>
```

Comment:
According to the [`Visitor::visit_string`](https://docs.serde.rs/serde/de/trait.Visitor.html#method.visit_string) documentation:

> It is never correct to implement `visit_string` without implementing `visit_str`.

In this case I think you want just `visit_str` because the implementation does not benefit from taking ownership of the String data.

Reply:
Fair point.